datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -2799,14 +2817,71 @@ impl PartialOrd for Extension {
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Hash)]
 pub struct Limit {
     /// Number of rows to skip before fetch
-    pub skip: usize,
+    pub skip: Option<Box<Expr>>,
     /// Maximum number of rows to fetch,
     /// None means fetching all rows
-    pub fetch: Option<usize>,
+    pub fetch: Option<Box<Expr>>,
     /// The logical plan
     pub input: Arc<LogicalPlan>,
 }
 
+/// Different types of skip expression in Limit plan.
+pub enum SkipType {
+    /// The skip expression is a literal value.
+    Literal(usize),
+    /// Currently only supports expressions that can be folded into constants.
+    UnsupportedExpr,
```

Comment:
Limit is a relational operator, so this will always need to be constant-foldable.

Reply:
Yes, but some expressions like `limit (select count(*) from t)` can't be folded at the planning stage , so we need to keep them for conversion into physical expressions later.

Reply:
Constant folding requires the `TypeCoercion` and `SimplifyExpressions` rule. We can't directly invoke them during planning for now and need to defer them to optimizer. 

Another reason the Limit node needs to contain expressions is to support Prepare statements. Issue #12294 requires the Limit node to use `Expr::Placeholder`.